### DISCUSSION TITLE
How to paginate with only a "page" parameter? [AGAIN] but with cache this time.

### DISCUSSION THREAD

**Discussion by GollyJer** [urql-graphql/urql#1760] [Q&A] [answered]
I [asked something similar before](https://github.com/FormidableLabs/urql/discussions/1719) but am completely stumped again.
I'm willing to pay someone to help at this point....

The following paging query hook looks a little ridiculous but the basic idea is 
" page until reaching a result that is smaller than `limit` "

It works **great** for `network-only`. See console log below.
```javascript 
export const useTransfers = ({
  chainId,
  limit = 100,
  recipient,
  sender,
  shouldPause = false,
  startPage = 1,
  testId = '',
}) => {
  const pause = !recipient || !sender || shouldPause;

  const page = useRef(startPage);
  const aggregateResults = useRef([]);
  const finalResult = useRef([]);
  const isDone = useRef(false);

  const prevSender = usePrevious(sender);
  if (sender !== prevSender) {
    page.current = 1;
    aggregateResults.current = [];
    finalResult.current = [];
    isDone.current = false;
  }

  const [, rerender] = useState({});

  const [query] = useQuery({
    query: GET_TRANSFERS,
    pause,
    requestPolicy: 'network-only',
    variables: {
      chainId,
      limit,
      page: page.current,
      recipient,
      sender,
    },
  });
  if (query.error) console.log('useTransfers ERROR\n', JSON.stringify(query.error), '\n'); // eslint-disable-line
  const queryResults = query.data?.transfers;

  useEffect(() => {
    const isInitialRender = prevSender === undefined;
    const hasResults = queryResults?.length > 0;
    const moduloResultsRemainder = queryResults?.length % limit;
    const isDonePaging = moduloResultsRemainder > 0;

    if (testId === 'to')
      console.log(
        'isInitialRender:',
        isInitialRender,
        ' | page',
        page.current,
        'isDonePaging:',
        isDonePaging,
        ' | hasResults:',
        hasResults,
        `(${queryResults?.length})`,
        ' | moduloResultsRemainder:',
        moduloResultsRemainder
      );

    if (!isInitialRender && hasResults && !isDonePaging) {
      page.current += 1;
      if (testId === 'to') console.log('query for page:', page.current);
      rerender({});
    }

    if (hasResults) {
      aggregateResults.current = [...aggregateResults.current, ...queryResults];
    }

    if (isDonePaging) {
      // dedupe for cache-and-network.
      finalResult.current = [
        ...aggregateResults.current.reduce((map, obj) => map.set(obj.id, obj), new Map()).values(),
      ];
      isDone.current = true;

      if (testId === 'to') console.log('done', finalResult.current?.length);
      rerender({});
    }
  }, [queryResults]); // 👈 Doesn't trigger when cache-only?

  return [finalResult.current, isDone.current];
};
```

Here are the results.
![image](https://user-images.githubusercontent.com/689204/124213217-0992b280-daa5-11eb-911b-1cecba295691.png)

But when set to `cache-first`, `cache-only` or `cache-and-network` it gets stuck.
![image](https://user-images.githubusercontent.com/689204/124213289-2c24cb80-daa5-11eb-8aac-2f54e448232f.png)

It seems the useEffect dependency, `queryResults`, doesn't update after forcing a rerender when querying the cache.

What am I missing here?
Thank you! 🙏💪

**Comment by JoviDeCroock** [comment, marked as answer]
```js
export const useTransfers = ({
  chainId,
  limit = 100,
  recipient,
  sender,
  shouldPause = false,
  startPage = 1,
  testId = '',
}) => {
  const pause = !recipient || !sender || shouldPause;

  const prevSender = usePrevious(sender);

  const [done, setDone] = useState(false);
  const [page, setPage] = useState(1);
  const [data, setData] = useState([]);

  const [query] = useQuery({
    query: GET_TRANSFERS,
    pause,
    variables: {
      chainId,
      limit,
      page,
      recipient,
      sender,
    },
  });

  const queryResults = query.data?.transfers;

  useEffect(() => {
    if (query.fetching || query.error) return
    const isInitialRender = prevSender === undefined;
    const hasResults = queryResults?.length > 0;
    const moduloResultsRemainder = queryResults?.length % limit;
    const isDonePaging = moduloResultsRemainder > 0;

    if (!isInitialRender && hasResults && !isDonePaging) {
      setPage(page + 1);
    }

    if (hasResults) {
      setData(d => [...d, ...queryResults]);
    }

    if (isDonePaging) {
      setDone(true);
    }
  }, [queryResults]);

  return [data, done];
};
```

**Reply by GollyJer** [reply]
This seems like it should work and is very similar to my initial try. 
However, it returns 0 results from the cache and never returns from the network. 😥
The code I posted very specifically controls when to render with the new page count so this doesn't happen.

 Sorry, I don't mean to waste your time. This has turned into a nightmare of a problem for our app.

**Reply by JoviDeCroock** [reply]
I mean, the only way this wouldn't return results if `pause: true` or if your request errors out or something else is going on. This code-piece you provide does not indicate an issue with the fetching-flow it was more of a data-flow one 😅 if `network-only` works then `cache-first` will work as well unless you have something going on in your caching layer that is preventing a forward

**Reply by GollyJer** [reply]
Yeah. That is why I was testing with `cache-only` to isolate that issue. You would think if `network-only` worked then `cache-only` would behave identically.

I guess it's possible the second page of results isn't in the cache?
It's tough debug the cache because we use Expo and the [urql devtools don't work with Expo managed apps](https://github.com/FormidableLabs/urql-devtools/issues/324).

**Reply by GollyJer** [reply]
Not the problem. I just tested `cache-only` starting on page 2 and correctly get 18 results returned.

So strange! 😕

**Reply by JoviDeCroock** [reply]
I mean, you can just use the `debugExchange` which is exported from `urql` as well, you can start out with just removing all your exchanges but the `fetchExchange` and that should make clear to you if your issue resides in React-land or Exchange-land.

Basically the idea being that the issue we're talking about is probably passing in `pause: true` and the solution provided above should give you the results and a done: true. If there's more things that come up as problematic feel free to comment further but afaict in relation to this specific piece of code it seems good.
